### PR TITLE
New Elastic Beanstalk deployment with /api/image support

### PR DESCRIPTION
**Pull request by kimyx**
Backend changes for

- single-instance elasticbeanstalk (no load balancer charges)
- development venv stored outside of the repo workspace
- EB-deployable matlab engine
- EB-deployable vector-code submodule
- support for new /api/image endpoint

### CODE REVIEW

File: application.py [hunk 1]
```diff
@@ -38,7 +38,7 @@ def call_matlab(d):
                    d["speed"], composition["o"], composition["o2"],
                    composition["n2"], composition["he"], composition["h"],
                    ACCOM_MODEL_DICT[d["accommodationModel"]],
-                   d["energyAccommodation"], d["surfaceMass"], [[]], "",
+                   d["energyAccommodation"], d["surfaceMass"], matlab.double([]), "",
```
Comment (kimyx) [inline application.py:41]: this allows singlepoint api to work without changing MAIN.m and prepares for multipoint.

File: application.py [hunk 2]
```diff
@@ -77,29 +77,47 @@ def allowed_file(filename):
...
-            image_fname = call_matlab_image(f.name)
-            return send_file(image_fname, mimetype='image/png')
+    # the file collection is lost when requests go through the API Gateway http proxy
+    file = request.files.get('file') or request.environ.get('wsgi.input')
+    if not file:
+        abort(411, f"file or wsgi.input argument is required\n{pprint.pformat(('REQUEST', request.environ))}")
+
+    # get the file path if possible
+    # it works for local flask mode and for direct access to EB instance
+    # it doesn't work for API Gateway http proxy in front of EB instance
```
Comment (kimyx) [inline application.py:87]: it does work with the API Gateway rest proxy, though.

File: application.py [hunk 3]
```diff
@@ -77,29 +77,47 @@ def allowed_file(filename):
...
+        
+        image_fname = call_matlab_image(f.name)
+        return send_file(image_fname, mimetype='image/png')
 
     # If we made it here, there was an unknown error
     abort(400, "Unknown error handling the uploaded image")
 
 
+@application.errorhandler(Exception)
+def handle_error(error):
```
Comment (kimyx) [inline application.py:111]: this provides a lot of information in the response when something fails in the back end.

File: application.py (hunk 1 above)
Comment (greglucas) [reply application.py:41]: Excellent find!

File: application.py (hunk 2 above)
Comment (greglucas) [reply application.py:87]: Interesting, was this the http Gateway that I set up before? That one never worked if that was the case, just me testing things out. I should just delete that API Gateway...

**Review by greglucas — APPROVED** [review]
This looks great! It looks like you've gone through a lot of testing with many different areas of the code.  Thanks for adding the tests in as well, I think that will be valuable going forward.

The private VPC integration can be done in a separate PR.

File: application.py (hunk 2 above)
Comment (kimyx) [reply application.py:87]: yes, same http gateway you set up before. It seems that it loses some part of the request while being passed through.

I read that the http gateway costs 1/3 as much as the rest gateway and is faster. It works fully with the changes here, although the file extension checking is circumvented. Actually parsing wrl files will confirm they are correct, so I don't think the file extension check is essential. I'm not sure if the http gateway supports private EC2s yet, though.

**Pull request merged**